src/neo/Network/P2P/TaskManager.cs

```diff
@@ -36,94 +38,65 @@ private class Timer { }
         /// </summary>
         private readonly HashSetCache<UInt256> knownHashes;
         private readonly Dictionary<UInt256, int> globalTasks = new Dictionary<UInt256, int>();
-        private readonly Dictionary<uint, TaskSession> receivedBlockIndex = new Dictionary<uint, TaskSession>();
-        private readonly HashSet<uint> failedSyncTasks = new HashSet<uint>();
         private readonly Dictionary<IActorRef, TaskSession> sessions = new Dictionary<IActorRef, TaskSession>();
         private readonly ICancelable timer = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(TimerInterval, TimerInterval, Context.Self, new Timer(), ActorRefs.NoSender);
-        private uint lastTaskIndex = 0;
+        private bool HasHeaderTask => globalTasks.ContainsKey(HeaderTaskHash);
 
         public TaskManager(NeoSystem system)
         {
             this.system = system;
             this.knownHashes = new HashSetCache<UInt256>(Blockchain.Singleton.MemPool.Capacity * 2 / 5);
-            this.lastTaskIndex = NativeContract.Ledger.CurrentIndex(Blockchain.Singleton.View);
             Context.System.EventStream.Subscribe(Self, typeof(Blockchain.PersistCompleted));
             Context.System.EventStream.Subscribe(Self, typeof(Blockchain.RelayResult));
         }
 
-        private bool AssignSyncTask(uint index, TaskSession filterSession = null)
+        private void OnHeaderTaskCompleted()
         {
-            if (index <= NativeContract.Ledger.CurrentIndex(Blockchain.Singleton.View) || sessions.Values.Any(p => p != filterSession && p.IndexTasks.ContainsKey(index)))
-                return true;
-            Random rand = new Random();
-            KeyValuePair<IActorRef, TaskSession> remoteNode = sessions.Where(p => p.Value != filterSession && p.Value.LastBlockIndex >= index)
-                .OrderBy(p => p.Value.IndexTasks.Count)
-                .ThenBy(s => rand.Next())
-                .FirstOrDefault();
-            if (remoteNode.Value == null)
-            {
-                failedSyncTasks.Add(index);
-                return false;
-            }
-            TaskSession session = remoteNode.Value;
-            session.IndexTasks.TryAdd(index, TimeProvider.Current.UtcNow);
-            remoteNode.Key.Tell(Message.Create(MessageCommand.GetBlockByIndex, GetBlockByIndexPayload.Create(index, 1)));
-            failedSyncTasks.Remove(index);
-            return true;
-        }
-
-        private void OnBlock(Block block)
-        {
-            var session = sessions.Values.FirstOrDefault(p => p.IndexTasks.ContainsKey(block.Index));
-            if (session is null) return;
-            session.IndexTasks.Remove(block.Index);
-            receivedBlockIndex.TryAdd(block.Index, session);
-            RequestTasks(false);
-        }
```

Comment:
Why this is removed?

Reply:
Have restored

Reply:
I think it's not correct. We should recover the `IndexTasks` too.

Reply:
@erikzhang Please have a check